File: app/src/main/java/com/diamond/appcliente/actividades/SubirComprobanteActivity.java
package com.diamond.appcliente.actividades;

import android.content.Intent;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.lifecycle.ViewModelProvider;
import com.diamond.appcliente.R;
import com.diamond.appcliente.viewmodel.ListarReservaViewModel;
import com.google.android.material.bottomnavigation.BottomNavigationView;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;

public class SubirComprobanteActivity extends AppCompatActivity {
    private static final int PICK_IMAGE_REQUEST = 1;
    private static final int REQUEST_CODE_PERMISSIONS = 1001;
    private ImageView imagePreview;
    private Uri imagenSeleccionadaUri;
    private ListarReservaViewModel listarReservaViewModel;
    private Long reservaId;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_subir_comprobante);
        this.reservaId = Long.valueOf(getIntent().getLongExtra("reservaId", -1));
        this.listarReservaViewModel = new ViewModelProvider(this).get(ListarReservaViewModel.class);

        Button btnSeleccionarImagen = findViewById(R.id.btnSeleccionarImagen);
        Button btnSubirComprobante = findViewById(R.id.btnSubirComprobante);
        Button btnCancelar = findViewById(R.id.btnCancelar);
        this.imagePreview = findViewById(R.id.imagePreview);

        // Request permissions if needed
        if (ContextCompat.checkSelfPermission(this, "android.permission.READ_EXTERNAL_STORAGE") != 0) {
            ActivityCompat.requestPermissions(this, new String[]{"android.permission.READ_EXTERNAL_STORAGE"}, REQUEST_CODE_PERMISSIONS);
        }

        getWindow().setStatusBarColor(-16777216);

        btnSeleccionarImagen.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                seleccionarImagen();
            }
        });

        btnSubirComprobante.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (imagenSeleccionadaUri != null) {
                    subirComprobante(reservaId, imagenSeleccionadaUri);
                } else {
                    Toast.makeText(SubirComprobanteActivity.this, "Por favor, seleccione una imagen", Toast.LENGTH_SHORT).show();
                }
            }
        });

        btnCancelar.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(SubirComprobanteActivity.this, ListarReservaActivity.class));
                finish();
            }
        });

        BottomNavigationView bottomNavigationView = findViewById(R.id.bottomNavigationView);
        bottomNavigationView.setSelectedItemId(R.id.historial);
        bottomNavigationView.setOnItemSelectedListener(new BottomNavigationView.OnItemSelectedListener() {
            @Override
            public boolean onNavigationItemSelected(MenuItem item) {
                int id = item.getItemId();
                if (id == R.id.nav_home) {
                    startActivity(new Intent(SubirComprobanteActivity.this, ClienteHomeActivity.class));
                    overridePendingTransition(0, 0);
                    return true;
                } else if (id == R.id.nav_servicios) {
                    startActivity(new Intent(SubirComprobanteActivity.this, SeccionServiciosActivity.class));
                    overridePendingTransition(0, 0);
                    return true;
                } else if (id == R.id.historial || id != R.id.nav_perfil) {
                    return false;
                } else {
                    startActivity(new Intent(SubirComprobanteActivity.this, UsuarioActivity.class));
                    overridePendingTransition(0, 0);
                    return true;
                }
            }
        });
    }

    // Request permission result
    @Override
    public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == REQUEST_CODE_PERMISSIONS && grantResults.length > 0 && grantResults[0] == 0) {
            Toast.makeText(this, "Permiso concedido", Toast.LENGTH_SHORT).show();
        }
    }

    // Seleccionar imagen de la galería
    private void seleccionarImagen() {
        Intent intent = new Intent(Intent.ACTION_PICK, MediaStore.Images.Media.EXTERNAL_CONTENT_URI);
        intent.setType("image/*");
        startActivityForResult(intent, PICK_IMAGE_REQUEST);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (resultCode == RESULT_OK && requestCode == PICK_IMAGE_REQUEST && data != null) {
            imagenSeleccionadaUri = data.getData();
            mostrarVistaPrevia(imagenSeleccionadaUri);
        }
    }

    private void mostrarVistaPrevia(Uri imagenUri) {
        try {
            imagePreview.setImageBitmap(BitmapFactory.decodeStream(getContentResolver().openInputStream(imagenUri)));
            imagePreview.setVisibility(View.VISIBLE);
        } catch (IOException e) {
            e.printStackTrace();
            Log.e("SubirComprobanteActivity", "Error al mostrar la vista previa de la imagen", e);
        }
    }

    // Subir el comprobante a la nube o servidor
    private void subirComprobante(Long reservaId2, Uri imagenUri) {
        try {
            InputStream inputStream = getContentResolver().openInputStream(imagenUri);
            File file = new File(getCacheDir(), "user_image.jpg");
            OutputStream outputStream = new FileOutputStream(file);
            byte[] buffer = new byte[1024];
            while (true) {
                int read = inputStream.read(buffer);
                if (read == -1) {
                    break;
                }
                outputStream.write(buffer, 0, read);
            }
            inputStream.close();
            outputStream.close();

            // Enviar la imagen al servidor
            listarReservaViewModel.subirComprobante(reservaId2, MultipartBody.Part.createFormData("imagen", file.getName(), RequestBody.create(MediaType.parse("image/*"), file)), new ListarReservaViewModel.ActualizarCallback() {
                @Override
                public void onSuccess(String mensaje) {
                    Log.d("SubirComprobanteActivity", "Comprobante subido exitosamente.");
                    Toast.makeText(SubirComprobanteActivity.this, mensaje, Toast.LENGTH_SHORT).show();
                    finish();
                }

                @Override
                public void onError(String mensaje) {
                    Log.e("SubirComprobanteActivity", "Error al subir el comprobante: " + mensaje);
                    Toast.makeText(SubirComprobanteActivity.this, mensaje, Toast.LENGTH_SHORT).show();
                }
            });
        } catch (IOException e) {
            e.printStackTrace();
            Log.e("SubirComprobanteActivity", "Error al manejar la imagen", e);
            Toast.makeText(this, "Error al manejar la imagen", Toast.LENGTH_SHORT).show();
        }
    }
}

File: app/src/main/java/com/diamond/appcliente/actividades/SubirComprobanteActivity.kt
package com.diamond.appcliente.actividades

import android.content.Intent
import android.graphics.BitmapFactory
import android.net.Uri
import android.os.Bundle
import android.provider.MediaStore
import android.util.Log
import android.view.View
import android.widget.Button
import android.widget.ImageView
import android.widget.Toast
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.diamond.appcliente.R
import com.diamond.appcliente.viewmodel.ListarReservaViewModel
import com.google.android.material.bottomnavigation.BottomNavigationView
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch
import okhttp3.MediaType
import okhttp3.MultipartBody
import okhttp3.RequestBody
import java.io.File
import java.io.IOException

@AndroidEntryPoint
class SubirComprobanteActivity : AppCompatActivity() {

    private lateinit var imagePreview: ImageView
    private var imagenSeleccionadaUri: Uri? = null
    private val listarReservaViewModel: ListarReservaViewModel by viewModels()
    private var reservaId: Long = -1L

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_subir_comprobante)

        reservaId = intent.getLongExtra("reservaId", -1L)
        imagePreview = findViewById(R.id.imagePreview)

        if (ContextCompat.checkSelfPermission(this, "android.permission.READ_EXTERNAL_STORAGE") != 0) {
            ActivityCompat.requestPermissions(this, arrayOf("android.permission.READ_EXTERNAL_STORAGE"), REQUEST_CODE_PERMISSIONS)
        }

        window.statusBarColor = -16777216

        findViewById<Button>(R.id.btnSeleccionarImagen).setOnClickListener { seleccionarImagen() }

        findViewById<Button>(R.id.btnSubirComprobante).setOnClickListener {
            if (imagenSeleccionadaUri != null) {
                subirComprobante(reservaId, imagenSeleccionadaUri!!)
            } else {
                Toast.makeText(this, "Por favor, seleccione una imagen", Toast.LENGTH_SHORT).show()
            }
        }

        findViewById<Button>(R.id.btnCancelar).setOnClickListener {
            startActivity(Intent(this, ListarReservaActivity::class.java))
            finish()
        }

        val bottomNavigationView = findViewById<BottomNavigationView>(R.id.bottomNavigationView)
        bottomNavigationView.selectedItemId = R.id.historial
        bottomNavigationView.setOnItemSelectedListener { item ->
            when (item.itemId) {
                R.id.nav_home -> { startActivity(Intent(this, ClienteHomeActivity::class.java)); overridePendingTransition(0, 0); true }
                R.id.nav_servicios -> { startActivity(Intent(this, SeccionServiciosActivity::class.java)); overridePendingTransition(0, 0); true }
                R.id.nav_perfil -> { startActivity(Intent(this, UsuarioActivity::class.java)); overridePendingTransition(0, 0); true }
                else -> false
            }
        }

        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                launch {
                    listarReservaViewModel.comprobanteEvento.collect { msg ->
                        Log.d("SubirComprobanteActivity", "Comprobante subido exitosamente.")
                        Toast.makeText(this@SubirComprobanteActivity, msg, Toast.LENGTH_SHORT).show()
                        finish()
                    }
                }
                launch {
                    listarReservaViewModel.error.collect { msg ->
                        Log.e("SubirComprobanteActivity", "Error al subir el comprobante: $msg")
                        Toast.makeText(this@SubirComprobanteActivity, msg, Toast.LENGTH_SHORT).show()
                    }
                }
            }
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == REQUEST_CODE_PERMISSIONS && grantResults.isNotEmpty() && grantResults[0] == 0) {
            Toast.makeText(this, "Permiso concedido", Toast.LENGTH_SHORT).show()
        }
    }

    private fun seleccionarImagen() {
        val intent = Intent(Intent.ACTION_PICK, MediaStore.Images.Media.EXTERNAL_CONTENT_URI)
        intent.type = "image/*"
        startActivityForResult(intent, PICK_IMAGE_REQUEST)
    }

    @Deprecated("Deprecated in Java")
    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode == RESULT_OK && requestCode == PICK_IMAGE_REQUEST && data != null) {
            imagenSeleccionadaUri = data.data
            mostrarVistaPrevia(imagenSeleccionadaUri!!)
        }
    }

    private fun mostrarVistaPrevia(imagenUri: Uri) {
        try {
            imagePreview.setImageBitmap(BitmapFactory.decodeStream(contentResolver.openInputStream(imagenUri)))
            imagePreview.visibility = View.VISIBLE
        } catch (e: IOException) {
            Log.e("SubirComprobanteActivity", "Error al mostrar la vista previa de la imagen", e)
        }
    }

    private fun subirComprobante(reservaId: Long, imagenUri: Uri) {
        try {
            val inputStream = contentResolver.openInputStream(imagenUri)!!
            val file = File(cacheDir, "user_image.jpg")
            file.outputStream().use { inputStream.copyTo(it) }
            inputStream.close()

            val imagenPart = MultipartBody.Part.createFormData(
                "imagen", file.name,
                RequestBody.create(MediaType.parse("image/*"), file)
            )
            listarReservaViewModel.subirComprobante(reservaId, imagenPart)
        } catch (e: IOException) {
            Log.e("SubirComprobanteActivity", "Error al manejar la imagen", e)
            Toast.makeText(this, "Error al manejar la imagen", Toast.LENGTH_SHORT).show()
        }
    }

    companion object {
        private const val PICK_IMAGE_REQUEST = 1
        private const val REQUEST_CODE_PERMISSIONS = 1001
    }
}
